src/hotspot/share/opto/doCall.cpp

```diff
@@ -1097,59 +1097,47 @@ ciMethod* Compile::optimize_inlining(ciMethod* caller, ciInstanceKlass* klass, c
+      callee->name() != ciSymbols::finalize_method_name()) {
+    return callee;
+  }
 
-    // All other interesting cases are instance klasses.
-    if (!receiver_type->isa_instptr()) {
-      return NULL;
-    }
+  // All other interesting cases are instance klasses.
+  if (!receiver_type->isa_instptr()) {
+    return NULL;
+  }
 
-    ciInstanceKlass *ikl = receiver_type->klass()->as_instance_klass();
-    if (ikl->is_loaded() && ikl->is_initialized() && !ikl->is_interface() &&
-        (ikl == actual_receiver || ikl->is_subtype_of(actual_receiver))) {
-      // ikl is a same or better type than the original actual_receiver,
-      // e.g. static receiver from bytecodes.
-      actual_receiver = ikl;
-      // Is the actual_receiver exact?
-      actual_receiver_is_exact = receiver_type->klass_is_exact();
-    }
+  ciInstanceKlass *ikl = receiver_type->klass()->as_instance_klass();
+  if (ikl->is_loaded() && ikl->is_initialized() && !ikl->is_interface() &&
+      (ikl == actual_receiver || ikl->is_subtype_of(actual_receiver))) {
+    // ikl is a same or better type than the original actual_receiver,
+    // e.g. static receiver from bytecodes.
+    actual_receiver = ikl;
+    // Is the actual_receiver exact?
+    actual_receiver_is_exact = receiver_type->klass_is_exact();
   }
 
   ciInstanceKlass*   calling_klass = caller->holder();
   ciMethod* cha_monomorphic_target = callee->find_monomorphic_target(calling_klass, klass, actual_receiver, check_access);
+
+  // Validate receiver info against target method.
   if (cha_monomorphic_target != NULL) {
-    assert(!cha_monomorphic_target->is_abstract(), "");
-    // Look at the method-receiver type.  Does it add "too much information"?
-    ciKlass*    mr_klass = cha_monomorphic_target->holder();
-    const Type* mr_type  = TypeInstPtr::make(TypePtr::BotPTR, mr_klass);
-    if (receiver_type == NULL || !receiver_type->higher_equal(mr_type)) {
-      // Calling this method would include an implicit cast to its holder.
-      // %%% Not yet implemented.  Would throw minor asserts at present.
-      // %%% The most common wins are already gained by +UseUniqueSubclasses.
-      // To fix, put the higher_equal check at the call of this routine,
-      // and add a CheckCastPP to the receiver.
-      if (TraceDependencies) {
-        tty->print_cr("found unique CHA method, but could not cast up");
-        tty->print("  method  = ");
-        cha_monomorphic_target->print();
-        tty->cr();
-      }
-      if (log() != NULL) {
-        log()->elem("missed_CHA_opportunity klass='%d' method='%d'",
-                       log()->identify(klass),
-                       log()->identify(cha_monomorphic_target));
+    bool has_receiver = !cha_monomorphic_target->is_static();
+    bool is_interface_holder = cha_monomorphic_target->holder()->is_interface();
+    if (has_receiver && !is_interface_holder) {
+      if (!cha_monomorphic_target->holder()->is_subtype_of(receiver_type->klass())) {
```

Comment:
Given we can't trust interface types, shouldn't we test for receiver_type not an interface here?

Reply:
Though it doesn't happen in practice now, we still can trust interface types on the receiver object because virtual/interface method invocation implies proper type check of the receiver.

Reply:
I think you need a CheckCastPP as the comment (written by me long ago IIRC) suggests.

I’d like to believe you can trust the type of a CHA receiver, but I don’t.  What the IR says about interface types is (last time I looked) just as untrustworthy as what the verifier thinks about them.

(We should fix this decisively someday.  Or did we do this cleanup already and missed it?  We have had dozens of bugs
in early C2 because of untrustworthy interface types.)

Suppose CHA (of default methods, yay!) says that since x is interface Foo, then default Foo::m is the only possible target
of x.m.  Before we execute Foo::m we still need a type check of x, because it could be any Object whatsoever.

I suggest adding the CheckCastPP in, in all modes, with fallback to de-opt not VM crash.  The IR types are saying there is not a dominating check.

Reply:
I closely looked at the relevant logic and couldn't find any possibility to bypass receiver subtype check. 

There are 2 call site shapes possible:

1. `invokevirtual C.m RECV => J.m DEFAULT` (`J.m` - resolved and selected method)

Verifier guarantees that RECV <: C (REFC) and method resolution ensures that C <: J (REFC <: DECC).


2. `invokeinterface I.m RECV => J.m DEFAULT` (`J.m` - resolved and selected method)

`I` (REFC) is required to be an interface (structural constraint of `invokeinterface` instruction).

Then, here's the relevant code:
```
  ciInstanceKlass* actual_receiver = resolved_iklass;
  ...
  ciInstanceKlass *ikl = receiver_type->klass()->as_instance_klass();
  if (ikl->is_loaded() && ikl->is_initialized() && !ikl->is_interface() &&
      (ikl == actual_receiver || ikl->is_subtype_of(actual_receiver))) {
    // ikl is a same or better type than the original actual_receiver,
    // e.g. static receiver from bytecodes.
    actual_receiver = ikl;
    // Is the actual_receiver exact?
    actual_receiver_is_exact = receiver_type->klass_is_exact();
  }

  ciInstanceKlass*   calling_klass = caller->holder();
  ciMethod* cha_monomorphic_target = resolved_method->find_monomorphic_target(calling_klass, resolved_iklass, actual_receiver, check_access);

...

ciMethod* ciMethod::find_monomorphic_target(...) {
  ...
  if (actual_recv->is_interface()) {
    // %%% We cannot trust interface types, yet.  See bug 6312651.
    return NULL;
  }

```

`receiver_type` can represent either an interface (1a) or a class (1b `ikl`).

1a. But if `ikl->is_interface()`, then `actual_receiver` stays as an interface (`I`, REFC) and `find_monomorphic_target()` returns NULL.

1b. If `ikl` is a class, then `ikl != actual_receiver` and depending on whether `ikl->is_subtype_of(actual_receiver))`, `actual_receiver` is either REFC (interface) or `receiver_type`.

`find_monomorphic_target()` can succeed only in the latter case (`actual_receiver` is a class and not an interface), but previous static subtype check guarantees that the receiver type (a class) implements the interface.  

Q.E.D

Reply:
So, CHA is still kept conservative when interface types on receiver are encountered. 
Currently, only single implementor case is optimised, but there are proper receiver subtype checks in place.

It should be quite straightforward to relax aforementioned constraints on receiver types (only classes, no interfaces), but then we would require additional receiver subtype check accompanying `Compile::optimize_virtual_call()` calls.

Reply:
Anyway, as a defense in depth measure, we could still insert receiver subtype check when CHA returns a default method and hope that the check will go away once type system better track interface types. Though the subtype check against an interface can be quite expensive, I don't expect any performance regressions considering CHA hasn't optimized default methods before.

Reply:
Evidently, it's not the whole story :-) Otherwise, we wouldn't have needed downcasts at all.

It becomes more interesting when CHA discovers unique concrete method under `REFC`:

3a. `invokevirtual   C.m RECV => J.m ABSTRACT, unique = K.m DEFAULT, K <: J` (`I.m` - resolved method, `REFC = C`)
3b. `invokeinterface I.m RECV => J.m ABSTRACT, unique = K.m DEFAULT, K <: J` (`I.m` - resolved method, `REFC = I`)

(Moreover, 3b is equivalent to 3a since `actual_receiver` has to be a class and
not an interface. Otherwise, CHA fails fast. There's additional case - unique
implementor - but it again strength-reduces invokeinterface to invokevirtual.)

It is guaranteed that `RECV <: C <: J` and `RECV <: I < J`, but there's nothing except the CHA itself that 
proves RECV <: J always.

In 3a and 3b case, CHA iterates all the subtypes under statically known receiver type
(`actual_receiver`, subtype of `I`/`C`). Assuming the analysis is correct, it proves
that `RECV <: J`.

Reply:
As a summary: if we feel safer when there's a dynamic receiver subtype check always present, it could be added. 

We would pay with performance (the check can become quite costly in some corner cases), but it shouldn't introduce performance regressions (compared to previous releases): CHA didn't return default methods before and interface call involves implicit receiver type check. 

(Need to be cautious not to replace a type profile-guided monomorphic case with a statically determined target interface method + interface subtype check. So, additional complexity in implementation as well.)